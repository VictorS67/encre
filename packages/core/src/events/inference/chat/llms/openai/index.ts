import {
  APIConnectionTimeoutError,
  APIUserAbortError,
  OpenAI as OpenAIClient,
} from 'openai';
import type { RequestOptions as OpenAIClientRequestOptions } from 'openai/core';

import { BaseLMCallOptions } from '../../base.js';

/**
 * Represents a call to a function as part of an OpenAI tool usage.
 * @example
 * ```typescript
 * const toolCall: OpenAIChatToolCall = {
 *   id: 'tool-id-from-api',
 *   type: 'function',
 *   function: {
 *     name: 'get_current_weather',
 *     arguments: "{\n\"location\": \"Boston, MA\"\n}"
 *   }
 * };
 * ```
 */
export interface OpenAIChatToolCall {
  /**
   * The ID of the tool call.
   */
  id: string;

  /**
   * The function that the model called.
   */
  function: {
    /**
     * The arguments to call the function with, as generated by the model in JSON
     * format. Note that the model does not always generate valid JSON, and may
     * hallucinate parameters not defined by your function schema. Validate the
     * arguments in your code before calling your function.
     */
    arguments: string;

    /**
     * The name of the function to call.
     */
    name: string;
  };

  /**
   * The type of the tool. Currently, only `function` is supported.
   */
  type: 'function';
}

/**
 * Specifies a tool the model should use. Use to force the model to call a specific
 * function.
 * @example
 * ```typescript
 * const toolChoice: OpenAIChatToolChoice = {
 *   type: 'function',
 *   function: {
 *     name: 'get_current_weather'
 *   }
 * };
 * ```
 */
export interface OpenAIChatToolChoice {
  /**
   * The type of the tool. Currently, only `function` is supported.
   */
  type: 'function';

  function: {
    /**
     * The name of the function to call.
     */
    name: string;
  };
}

/**
 * Represents a tool available for the model to use during processing.
 * @example
 * ```typescript
 * const tool: OpenAIChatTool = {
 *   type: 'function',
 *   function: {
 *     name: 'get_current_weather',
 *     description: 'get weather in a given location',
 *     parameters: {
 *       type: 'object',
 *       properties: {
 *         location: {
 *           type: 'string'
 *         },
 *         unit: {
 *           type: 'string',
 *           enum: ['celsius', 'fahrenheit'],
 *         }
 *       },
 *       required: ['location']
 *     }
 *   }
 * };
 * ```
 */
export interface OpenAIChatTool {
  /**
   * The type of the tool. Currently, only `function` is supported.
   */
  type: 'function';

  function: OpenAIFunctionDef;
}

/**
 * Details of a function that can be called by the model.
 * @example
 * ```typescript
 * const functionDef: OpenAIFunctionDef = {
 *   name: 'get_current_weather',
 *   description: 'get weather in a given location',
 *   parameters: {
 *     type: 'object',
 *     properties: {
 *       location: {
 *         type: 'string'
 *       },
 *       unit: {
 *         type: 'string',
 *         enum: ['celsius', 'fahrenheit'],
 *       }
 *     },
 *     required: ['location']
 *   }
 * };
 * ```
 */
export interface OpenAIFunctionDef {
  /**
   * The name of the function to be called. Must be a-z, A-Z, 0-9, or contain
   * underscores and dashes, with a maximum length of 64.
   */
  name: string;

  /**
   * A description of what the function does, used by the model to choose when and
   * how to call the function.
   */
  description?: string;

  /**
   * The parameters the functions accepts, described as a JSON Schema object. See the
   * [guide]({@link https://platform.openai.com/docs/guides/text-generation/function-calling})
   * for examples, and the
   * [JSON Schema reference]({@link https://json-schema.org/understanding-json-schema/}) for
   * documentation about the format.
   *
   * Omitting `parameters` defines a function with an empty parameter list.
   */
  parameters?: Record<string, unknown>;
}

/**
 * Enhanced options for OpenAI chat-related API calls.
 */
export interface OpenAIChatCallOptions extends OpenAICallOptions {
  /**
   * Whether to return log probabilities of the output tokens or not. If true,
   * returns the log probabilities of each output token returned in the content
   * of message. This option is currently not available on the `gpt-4-vision-preview` model.
   */
  logprobs?: boolean;

  /**
   * An integer between 0 and 20 specifying the number of most likely tokens to
   * return at each token position, each with an associated log probability.
   * `logprobs` must be set to `true` if this parameter is used.
   */
  topLogprobs?: number;

  /**
   * Controls which (if any) function is called by the model. `none` means the model
   * will not call a function and instead generates a message. `auto` means the model
   * can pick between generating a message or calling a function. Specifying a
   * particular function via
   * `{"type": "function", "function": {"name": "my_function"}}` forces the model to
   * call that function.
   *
   * `none` is the default when no functions are present. `auto` is the default if
   * functions are present.
   */
  toolChoice?: 'none' | 'auto' | OpenAIChatToolChoice;

  /**
   * A list of tools the model may call. Currently, only functions are supported as a
   * tool. Use this to provide a list of functions the model may generate JSON inputs
   * for. A max of 128 functions are supported.
   */
  tools?: Array<OpenAIChatTool>;
}

/**
 * Options specific to text-based API calls.
 */
export interface OpenAITextCallOptions extends OpenAICallOptions {
  /**
   * The suffix that comes after a completion of inserted text.
   *
   * This parameter is only supported for model `gpt-3.5-turbo-instruct`.
   */
  suffix?: string;
}

/**
 * Base options for any OpenAI API call.
 */
export interface OpenAICallOptions extends BaseLMCallOptions {
  /**
   * Additional options to pass to the underlying axios request.
   */
  options?: OpenAIClientRequestOptions;
}

/**
 * Core settings for interacting with OpenAI API.
 */
export interface OpenAIBaseInput {
  /**
   * ID of the model to use. You can use the [List models]({@link https://platform.openai.com/docs/api-reference/models/list})
   * API to see all of your available models, or see [Model overview]({@link https://platform.openai.com/docs/models/overview})
   * for descriptions of them.
   */
  modelName: string;

  /**
   * Number between -2.0 and 2.0. Positive values penalize new tokens based on their
   * existing frequency in the text so far, decreasing the model's likelihood to
   * repeat the same line verbatim.
   *
   * [See more information about frequency and presence panalties]({@link https://platform.openai.com/docs/guides/text-generation/parameter-details})
   */
  frequencyPenalty: number;

  /**
   * Number between -2.0 and 2.0. Positive values penalize new tokens based on
   * whether they appear in the text so far, increasing the model's likelihood to
   * talk about new topics.
   *
   * [See more information about frequency and presence penalties]({@link https://platform.openai.com/docs/guides/text-generation/parameter-details})
   */
  presencePenalty: number;

  /**
   * Whether to stream back partial progress. If set, tokens will be sent as
   * data-only [server-sent events]({@link https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#Event_stream_format})
   * as they become available, with the stream terminated by a `data: [DONE]`
   * message.
   *
   * [Example Python code]({@link https://cookbook.openai.com/examples/how_to_stream_completions})
   */
  streaming: boolean;

  /**
   * What sampling temperature to use, between 0 and 2. Higher values like 0.8 will
   * make the output more random, while lower values like 0.2 will make it more
   * focused and deterministic.
   *
   * We generally recommend altering this or `topP` but not both.
   */
  temperature: number;

  /**
   * An alternative to sampling with temperature, called nucleus sampling, where the
   * model considers the results of the tokens with temperature probability mass. So 0.1
   * means only the tokens comprising the top 10% probability mass are considered.
   *
   * We generally recommend altering this or `temperature` but not both.
   */
  topP: number;

  /**
   * How many completions to generate for each prompt.
   *
   * **Note:** Because this parameter generates many completions, it can quickly
   * consume your token quota. Use carefully and ensure that you have reasonable
   * settings for `maxTokens` and `stopWords`.
   */
  n: number;

  /**
   * Modify the likelihood of specified tokens appearing in the completion.
   *
   * Accepts a json object that maps tokens (specified by their token ID in the GPT
   * tokenizer) to an associated bias value from -100 to 100. You can use this
   * [tokenizer tool]({@link https://platform.openai.com/tokenizer?view=bpe})
   * (which works for both GPT-2 and GPT-3) to convert text to token IDs. Mathematically,
   * the bias is added to the logits generated by the model prior to sampling.
   * The exact effect will vary per model, but values between -1 and 1 should
   * decrease or increase likelihood of selection; values like -100 or 100 should
   * result in a ban or exclusive selection of the relevant token.
   *
   * As an example, you can pass `{"50256": -100}` to prevent the end-of-text token
   * from being generated.
   */
  logitBias?: Record<string, number>;

  /**
   * The maximum number of [tokens]({@link https://platform.openai.com/tokenizer})
   * to generate in the completion.
   *
   * The token count of your prompt plus `maxTokens` cannot exceed the model's
   * context length.
   *
   * [Example Python code]({@link https://cookbook.openai.com/examples/how_to_count_tokens_with_tiktoken}) f
   * or counting tokens.
   */
  maxTokens?: number;

  /**
   * This feature is in Beta. If specified, our system will make a best effort to
   * sample deterministically, such that repeated requests with the same `seed` and
   * parameters should return the same result. Determinism is not guaranteed, and you
   * should refer to the `system_fingerprint` response parameter to monitor changes
   * in the backend.
   */
  seed?: number;

  /**
   * Up to 4 sequences where the API will stop generating further tokens. The
   * returned text will not contain the stop sequence.
   */
  stopWords?: string[];

  /**
   * A unique identifier representing your end-user, which can help OpenAI to monitor
   * and detect abuse.
   *
   * [Learn more]({@link https://platform.openai.com/docs/guides/safety-best-practices})
   */
  user?: string;

  /**
   * Holds any additional parameters that are valid to pass to
   * [OpenAI API Reference]({@link https://platform.openai.com/docs/api-reference/introduction})
   * that are not explicitly specified on this class.
   */
  additionalKwargs?: Record<string, unknown>;

  /**
   * Timeout to use when making requests to OpenAI.
   */
  timeout?: number;

  /**
   * API key to use when making requests to OpenAI. Defaults to the value of
   * `OPENAI_API_KEY` environment variable.
   */
  openAIApiKey?: string;
}


/**
 * Configuration for standard OpenAI API input parameters.
 */
export interface OpenAIInput extends OpenAIBaseInput {
  /**
   * Generates `bestOf` completions server-side and returns the "best" (the one with
   * the highest log probability per token). Results cannot be streamed.
   *
   * When used with `n`, `bestOf` controls the number of candidate completions and
   * `n` specifies how many to return – `bestOf` must be greater than `n`.
   *
   * **Note:** Because this parameter generates many completions, it can quickly
   * consume your token quota. Use carefully and ensure that you have reasonable
   * settings for `maxToken` and `stopWords`.
   */
  bestOf?: number;

  /**
   * Include the log probabilities on the `logprobs` most likely tokens, as well the
   * chosen tokens. For example, if `logprobs` is 5, the API will return a list of
   * the 5 most likely tokens. The API will always return the `logprob` of the
   * sampled token, so there may be up to `logprobs+1` elements in the response.
   *
   * The maximum value for `logprobs` is 5.
   */
  logprobs?: number;

  /**
   * Echo back the prompt in addition to the completion
   */
  echo?: boolean;
}

/**
 * Configuration for chat-specific input settings in OpenAI API calls.
 */
export interface OpenAIChatInput extends OpenAIBaseInput {
  /**
   * An object specifying the format that the model must output. Compatible with
   * [GPT-4 Turbo](https://platform.openai.com/docs/models/gpt-4-and-gpt-4-turbo) and
   * all GPT-3.5 Turbo models newer than `gpt-3.5-turbo-1106`.
   *
   * Setting to `"json"` enables JSON mode, which guarantees the message the model
   * generates is valid JSON.
   *
   * **Important:** when using JSON mode, you **must** also instruct the model to
   * produce JSON yourself via a system or user message. Without this, the model may
   * generate an unending stream of whitespace until the generation reaches the token
   * limit, resulting in a long-running and seemingly "stuck" request. Also note that
   * the message content may be partially cut off if `finish_reason="length"`, which
   * indicates the generation exceeded `maxTokens` or the conversation exceeded the
   * max context length.
   */
  responseFormatType?: 'text' | 'json';

  /** ChatGPT messages to pass as a prefix to the prompt */
  chatMessages?: OpenAIClient.Chat.ChatCompletionMessageParam[];
}

/**
 * Checks if a model name is suitable for OpenAI chat functions.
 * @param modelName The model name to check.
 * @returns True if the model is compatible with chat functions.
 */
export function checkModelForOpenAIChat(modelName?: string): boolean {
  return (
    modelName !== undefined &&
    (modelName.startsWith('gpt-3.5-turbo') || modelName.startsWith('gpt-4')) &&
    !modelName.includes('-instruct')
  );
}

/**
 * Checks if a model name is suitable for OpenAI vision-related tasks.
 * @param modelName The model name to check.
 * @returns True if the model is suitable for vision tasks.
 */
export function checkModelForOpenAIVision(modelName?: string): boolean {
  return (
    modelName !== undefined &&
    modelName.startsWith('gpt-4') &&
    modelName.includes('vision')
  );
}

/**
 * Wraps errors from the OpenAI client to standardize error handling.
 * @param e The original error thrown by the OpenAI client.
 * @returns A standardized error object.
 */
export function wrapOpenAIClientError(e: Error): Error {
  let error: Error;
  if (e.constructor.name === APIConnectionTimeoutError.name) {
    error = new Error(e.message);
    error.name = 'TimeoutError';
  } else if (e.constructor.name === APIUserAbortError.name) {
    error = new Error(e.message);
    error.name = 'AbortError';
  } else {
    error = e;
  }

  return error;
}
