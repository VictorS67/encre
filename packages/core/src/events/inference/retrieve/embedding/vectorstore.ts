import { type Context } from '../../../input/load/docs/index.js';
import { type BaseVectorStore } from '../../../input/load/vectorstore/index.js';
import {
  BaseEmbeddingRetriever,
  type BaseEmbeddingRetrieverParams,
} from '../base.js';
import {
  type MaxMarginalRelevanceSearch,
  type SimilaritySearch,
  type VectorStoreRetrieverSearch,
} from './index.js';

/**
 * Parameters for initializing a VectorStoreRetriever, specifying the vector store,
 * optional filtering criteria, and search strategy.
 */
export interface VectorStoreRetrieverParams<
  V extends BaseVectorStore = BaseVectorStore,
> extends BaseEmbeddingRetrieverParams {
  /**
   * The vector store used for retrieving vectors.
   */
  vectorstore: BaseVectorStore;

  /**
   * Optional filtering type specific to the vector store.
   */
  filter?: V['FilterType'];

  /**
   * Search strategy to use, can be either similarity search or max marginal
   * relevance (MMR) search.
   */
  search?: VectorStoreRetrieverSearch;
}

/**
 * A retriever that utilizes a vector store to fetch contextually relevant data based
 * on the provided vector query. It supports configurable search strategies such as
 * similarity search and max marginal relevance (MMR).
 *
 * @template V - The type of the vector store, extending from `BaseVectorStore`.
 *
 * @example
 * ```
 * const vectorStore = new MemoryVectorStore();
 * const retriever = new VectorStoreRetriever<MemoryVectorStore>({
 *   vectorstore: vectorStore,
 *   search: {
 *     type: 'similarity',
 *     topK: 5
 *   }
 * });
 *
 * // To retrieve using a vector query
 * const results = await retriever.invoke([0.1, 0.2, 0.3, 0.4]);
 * console.log(results);
 * ```
 */
export class VectorStoreRetriever<V extends BaseVectorStore = BaseVectorStore>
  extends BaseEmbeddingRetriever
  implements VectorStoreRetrieverParams<V>
{
  _isSerializable = true;

  static _name(): string {
    return 'VectorStoreRetriever';
  }

  _retrieverType(): string {
    return 'vectorstore';
  }

  /**
   * The vector store used for retrieving vectors.
   */
  vectorstore: BaseVectorStore;

  /**
   * Optional filtering type specific to the vector store.
   */
  filter?: V['FilterType'];

  /**
   * Search strategy to use, can be either similarity search or max marginal
   * relevance (MMR) search.
   */
  search: VectorStoreRetrieverSearch;

  constructor(fields: VectorStoreRetrieverParams<V>) {
    const _fields = {
      search:
        fields.search?.type === 'mmr'
          ? ({
              type: 'mmr',
              topK: 4,
              lambda: 0.5,
            } as MaxMarginalRelevanceSearch)
          : ({
              type: 'similarity',
              topK: 4,
            } as SimilaritySearch),
      ...fields,
    };

    super(_fields);

    this.vectorstore = _fields.vectorstore;
    this.filter = _fields.filter;
    this.search = _fields.search;
  }

  /**
   * Retrieves contextually relevant data from the vector store based on the provided query vector.
   * The method of retrieval depends on the configured search type, which can be either a similarity search
   * or a max marginal relevance (MMR) search.
   *
   * @param {number[]} query - The query vector used to find similar vectors in the vector store.
   * @param {SerializedCallOptions} options - Additional serialized options that may affect retrieval, such as safety settings.
   * @returns {Promise<Context[]>} A promise that resolves to an array of contexts, which are the retrieved data elements based on the query.
   * @throws {Error} Throws an error if the search type is unsupported, indicating a configuration or implementation issue.
   * @internal
   * @example
   * Using Similarity Search:
   * ```typescript
   * // Assuming 'retriever' is an instance of VectorStoreRetriever and initialized properly
   * const queryVector = [0.1, 0.2, 0.3, 0.4];
   * retriever.search = {
   *   type: 'similarity',
   *   topK: 5
   * };
   * const results = await retriever._retrieve(queryVector, {});
   * console.log(results); // logs contexts similar to the query vector
   * ```
   *
   * Using Max Marginal Relevance (MMR) Search:
   * ```typescript
   * // Setting up the retriever for MMR search
   * retriever.search = {
   *   type: 'mmr',
   *   topK: 5,
   *   lambda: 0.65
   * };
   * const mmrResults = await retriever._retrieve(queryVector, {});
   * console.log(mmrResults); // logs diversified search results balancing similarity and diversity
   * ```
   *
   */
  async _retrieve(
    query: number[],
    options: this['SerializedCallOptions']
  ): Promise<Context[]> {
    if (this.search.type === 'mmr') {
      const searches = await this.vectorstore.maxMarginalRelevanceSearch(
        query,
        this.search.topK ?? 4,
        this.search.lambda ?? 0.5,
        this.filter
      );

      return searches.map((s) => s[0]);
    } else if (this.search.type === 'similarity') {
      const searches = await this.vectorstore.similaritySearch(
        query,
        this.search.topK ?? 4,
        this.filter
      );

      return searches.map((s) => s[0]);
    } else {
      throw new Error(
        `Does not support search type in vector store retriever: ${
          (this.search as any).type
        }`
      );
    }
  }
}
