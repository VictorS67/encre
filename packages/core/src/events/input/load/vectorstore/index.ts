export * from './base.js';
export * from './chroma.js';
export * from './memory.js';