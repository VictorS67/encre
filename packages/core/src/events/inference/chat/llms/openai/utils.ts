import { OpenAI as OpenAIClient } from 'openai';

import {
  type BaseMessage,
  type ContentLike,
  MessageRole,
  ChatMessage,
} from '../../../../input/load/msgs/index.js';

/**
 * Determines if a message requires multi-modal handling based on its content. Multi-modal
 * models do not support function/tool calls.
 *
 * @param message - The message to check for modality requirements.
 * @returns A boolean indicating if multi-modal processing is needed for the message.
 * @example
 * ```typescript
 * const message = new HumanMessage({ content: "Hello" });
 * const isMultiModalNeeded = isModalityRequiredInMessage(message);
 * console.log(isMultiModalNeeded); // Output: false
 * ```
 */
export function isModalityRequiredInMessage(message: BaseMessage): boolean {
  // Multi-modal model does not suport any function/tools
  if (message._role() === 'function') {
    return false;
  }

  const isModalityRequired = (content: ContentLike): boolean => {
    return typeof content !== 'string' && isChatCompletionContentPart(content);
  };

  if (!Array.isArray(message.content)) {
    return isModalityRequired(message.content);
  }

  return message.content.some((content: ContentLike) =>
    isModalityRequired(content)
  );
}

/**
 * Checks if a part of a message matches the expected structure of a chat completion content
 * part in OpenAI's API.
 *
 * @param partLike - The object to check.
 * @returns A boolean indicating if the object is a valid chat completion content part.
 */
export function isChatCompletionContentPart(
  partLike: any
): partLike is OpenAIClient.ChatCompletionContentPart {
  return (
    typeof partLike === 'object' &&
    (isChatCompletionContentPartImage(partLike) ||
      isChatCompletionContentPartText(partLike))
  );
}

/**
 * Determines if a content part is specifically an image, based on OpenAI's chat completion
 * content structure.
 *
 * @param partLike - The object to check.
 * @returns A boolean indicating if the part is an image content part.
 */
export function isChatCompletionContentPartImage(
  partLike: any
): partLike is OpenAIClient.ChatCompletionContentPartImage {
  return (
    typeof partLike === 'object' &&
    'type' in partLike &&
    partLike['type'] === 'image_url' &&
    'image_url' in partLike &&
    isOpenAIChatImageURL(partLike['image_url'])
  );
}

/**
 * Validates whether an object is structured as an OpenAI chat image URL, including checking
 * detail levels.
 *
 * @param imageUrlLike - The object to validate.
 * @returns A boolean indicating if the object is a valid OpenAI chat image URL.
 */
export function isOpenAIChatImageURL(
  imageUrlLike: any
): imageUrlLike is OpenAIClient.ChatCompletionContentPartImage.ImageURL {
  if (typeof imageUrlLike !== 'object') return false;

  if (
    'detail' in imageUrlLike &&
    !['auto', 'low', 'high'].includes(imageUrlLike['detail'])
  ) {
    return false;
  }

  return 'url' in imageUrlLike && typeof imageUrlLike['url'] === 'string';
}

/**
 * Verifies if a part of a message is structured as a text part, according to the OpenAI chat
 * API's expectations.
 *
 * @param partLike - The object to check.
 * @returns A boolean indicating if the part is a text content part.
 */
export function isChatCompletionContentPartText(
  partLike: any
): partLike is OpenAIClient.ChatCompletionContentPartText {
  return (
    typeof partLike === 'object' &&
    'type' in partLike &&
    partLike['type'] === 'text' &&
    'text' in partLike &&
    typeof partLike['text'] === 'string'
  );
}

/**
 * Checks if an object represents a tool call within a chat message, matching the structure
 * expected by the OpenAI API.
 *
 * @param toolLike - The object to check.
 * @returns A boolean indicating if the object is a valid tool call.
 */
export function isChatCompletionMessageToolCall(
  toolLike: any
): toolLike is OpenAIClient.ChatCompletionMessageToolCall {
  return (
    typeof toolLike === 'object' &&
    'type' in toolLike &&
    toolLike['type'] === 'function' &&
    'id' in toolLike &&
    typeof toolLike['id'] === 'string' &&
    'function' in toolLike &&
    isOpenAIFunction(toolLike['function'])
  );
}

/**
 * Validates whether an object is structured as a function call for OpenAI's chat API,
 * including argument validation.
 *
 * @param functionLike - The object to validate.
 * @returns A boolean indicating if the object is a valid function structure.
 */
export function isOpenAIFunction(
  functionLike: any
): functionLike is OpenAIClient.ChatCompletionMessageToolCall.Function {
  return (
    typeof functionLike === 'object' &&
    'name' in functionLike &&
    typeof functionLike['name'] === 'string' &&
    /**
     * The following description is the description of `arguments`:
     *
     * The arguments to call the function with, as generated by the model in JSON
     * format. Note that the model does not always generate valid JSON, and may
     * hallucinate parameters not defined by your function schema. Validate the
     * arguments in your code before calling your function.
     *
     * It seems that this parameter is containing a JSON object, even the type is string.
     * So we allow object type in this parameter, but we will convert it to string before
     * calling the model API.
     */
    'arguments' in functionLike &&
    (typeof functionLike['arguments'] === 'string' ||
      typeof functionLike['arguments'] === 'object')
  );
}

/**
 * Converts a BaseMessage into a format suitable for OpenAI's Chat API, handling different
 * roles such as system, human, and assistant.
 *
 * @param message - The message to convert.
 * @returns An ChatCompletionMessageParam object representing the converted message.
 * @example
 * ```
 * const message = new HumanMessage({ content: "Example content });
 * const contentParam = getContentFromMessage(message);
 * console.log(contentParam);
 * ```
 */
export function getContentFromMessage(
  message: BaseMessage
): OpenAIClient.Chat.ChatCompletionMessageParam {
  const role: MessageRole = message._role();

  const getContentWithRole = (
    message: BaseMessage,
    role: string
  ): OpenAIClient.Chat.ChatCompletionMessageParam => {
    if (role === 'system') {
      return {
        role,
        name: message.name,
        content: coerceMessageContentToString(message.content),
      };
    } else if (role === 'human') {
      return {
        role: 'user',
        name: message.name,
        content:
          typeof message.content === 'string'
            ? message.content
            : getPartsFromMessageContent(message.content),
      };
    } else if (role === 'assistant') {
      return {
        role,
        name: message.name,
        content: coerceMessageContentToString(message.content),
        tool_calls: getToolCallsFromMessageKwargs(message.additionalKwargs),
      };
    } else if (role === 'function') {
      return {
        role: 'tool',
        content: coerceMessageContentToString(message.content),
        tool_call_id: getToolCallIdFromMessageKwargs(message.additionalKwargs),
      };
    }

    throw new Error(`CANNOT get OpenAIChat content because of role: ${role}`);
  };

  if (role !== 'general') {
    return getContentWithRole(message, role);
  }

  const generalMessage = message as ChatMessage;
  if (
    ['human', 'assistant', 'system', 'function'].includes(generalMessage.role)
  ) {
    return getContentWithRole(generalMessage, generalMessage.role);
  }

  return getContentWithRole(generalMessage, 'human');
}

/**
 * Checks if a string content is in valid JSON format.
 *
 * @param content - The content string to check.
 * @returns A boolean indicating if the content is valid JSON.
 */
export function isJSONInContent(content: string): boolean {
  try {
    JSON.parse(content);
    return true;
  } catch (e) {
    console.warn('content is not JSON');
  }

  return false;
}

/**
 * Attempts to parse and reformat a JSON string into a more structured JSON object.
 *
 * @param content - The JSON string to parse and format.
 * @returns A JSON object if parsing is successful, or null if an error occurs.
 */
export function formatJSONInContent(
  content: string
): Record<string, unknown> | null {
  try {
    return JSON.parse(content);
  } catch (e) {
    console.warn('content is not JSON, cannot format');
  }

  return null;
}

/**
 * Attempts to reformat a JSON string for better readability or further processing.
 *
 * @param content - The JSON string to format.
 * @returns A formatted JSON string.
 */
export function formatJSONStringInContent(content: string): string {
  try {
    JSON.parse(content);
    return formatArguments(content);
  } catch (e) {
    console.warn('content is not JSON, cannot format arguments');
  }

  return content;
}

/**
 * Formats a JSON string representing arguments into a more readable and properly indented string format.
 * This function is useful when you need to display or log the structured arguments in a human-readable
 * form, especially for debugging or presentation purposes.
 *
 * The function parses the JSON string to an object, iterates over its properties, and constructs a
 * formatted string with each property on a new line and indented for clarity.
 *
 * @param args - A JSON string representing the arguments that need formatting.
 * @returns A string that represents the formatted arguments in a structured, pretty-printed format.
 * @example
 * ```typescript
 * const args = '{"name":"John Doe","age":30,"skills":["JavaScript", "TypeScript"]}';
 * const formatted = formatArguments(args);
 * console.log(formatted);
 * // Output:
 * // {
 * //   "name":"John Doe",
 * //   "age":30,
 * //   "skills":["JavaScript","TypeScript"]
 * // }
 * ```
 */
export function formatArguments(args: string): string {
  const lines: string[] = [];

  lines.push('{');
  const json = JSON.parse(args);

  const properties: string[] = [];
  for (const fieldName of Object.keys(json)) {
    properties.push(`"${fieldName}":${formatValue(json[fieldName])}`);
  }

  lines.push(properties.join(',\n'));
  lines.push('}');

  return lines.join('\n');
}

function formatValue(value: unknown): string {
  if (typeof value === 'string') {
    return `"${value}"`;
  }
  if (typeof value === 'number') {
    return `${value}`;
  }
  if (Array.isArray(value)) {
    return `[${value.map((v) => formatValue(v)).join(',')}]`;
  }
  return '""';
}

function coerceMessageContentToString(
  content: ContentLike | ContentLike[]
): string {
  const coerceSingleContentLikeToString = (content: ContentLike): string => {
    if (typeof content === 'string') {
      return content;
    }

    return JSON.stringify(content, null, 2);
  };

  if (Array.isArray(content)) {
    return content.map((c) => coerceSingleContentLikeToString(c)).join('\n\n');
  }

  return coerceSingleContentLikeToString(content);
}

function getPartsFromMessageContent(
  content: ContentLike | ContentLike[]
): Array<OpenAIClient.ChatCompletionContentPart> {
  const getPart = (
    content: ContentLike
  ): OpenAIClient.ChatCompletionContentPart => {
    if (typeof content === 'string') {
      return {
        type: 'text',
        text: content,
      };
    } else if (isChatCompletionContentPartText(content)) {
      return content;
    } else if (isChatCompletionContentPartImage(content)) {
      return content;
    }

    throw new Error('Message is not valid for OpenAIChat');
  };

  if (!Array.isArray(content)) {
    return [getPart(content)];
  }

  return content.map((content: ContentLike) => getPart(content));
}

function getToolCallsFromMessageKwargs(
  additionalKwargs: Record<string, unknown> | undefined
): Array<OpenAIClient.ChatCompletionMessageToolCall> {
  if (
    additionalKwargs === undefined ||
    !('tool_calls' in additionalKwargs) ||
    !Array.isArray(additionalKwargs['tool_calls'])
  ) {
    return [];
  }

  if (
    additionalKwargs['tool_calls'].every((tc) =>
      isChatCompletionMessageToolCall(tc)
    )
  ) {
    return additionalKwargs['tool_calls'].map((tc) => ({
      ...tc,
      function: {
        ...tc.function,
        arguments:
          typeof tc.function.arguments === 'object'
            ? JSON.stringify(tc.function.arguments, null, 2)
            : tc.function.arguments,
      },
    }));
  }

  throw new Error('Message kwarg `tool_calls` is not valid for OpenAIChat');
}

function getToolCallIdFromMessageKwargs(
  additionalKwargs: Record<string, unknown> | undefined
) {
  if (
    additionalKwargs === undefined ||
    !('tool_call_id' in additionalKwargs) ||
    !(typeof additionalKwargs['tool_call_id'] === 'string')
  ) {
    console.warn(
      'Function message does not have a `tool_call_id` in OpenAI ToolMessage'
    );

    return '';
  }

  return additionalKwargs['tool_call_id'];
}
