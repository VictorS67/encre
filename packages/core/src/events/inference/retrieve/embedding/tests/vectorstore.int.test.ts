import { describe, expect, test } from '@jest/globals';
import { OpenAIEmbeddings } from '../../../../embeddings/openai.js';
import { MemoryVectorStore } from '../../../../input/load/vectorstore/memory.js';
import { VectorStoreRetriever } from '../vectorstore.js';

describe('VectorStoreRetriever', () => {
  const OPENAI_API_KEY = 'you_should_get_this_api_from_openai';

  const openaiEmbeddings = new OpenAIEmbeddings({
    openAIApiKey: OPENAI_API_KEY,
    modelName: 'text-embedding-ada-002',
  });

  const embed = async (query: string): Promise<number[]> => {
    const embedResult = await openaiEmbeddings.invoke(query);

    return embedResult.embedding;
  };

  test('similarity search', async () => {
    const memoryVectorStore = new MemoryVectorStore({
      vectors: [
        {
          id: 'vec1',
          content: 'I like pizza',
          embedding: await embed('I like pizza'),
          metadata: {},
        },
        {
          id: 'vec2',
          content: "I don't like dogs",
          embedding: await embed("I don't like dogs"),
          metadata: {},
        },
        {
          id: 'vec2',
          content: 'I looks like a clown',
          embedding: await embed('I looks like a clown'),
          metadata: {},
        },
      ],
    });

    const retriever = new VectorStoreRetriever({
      vectorstore: memoryVectorStore,
      search: {
        type: 'similarity',
        topK: 2,
      },
    });

    expect(
      await retriever.invoke(await embed('what do I like?'))
    ).toMatchSnapshot();
  });

  test('mmr search', async () => {
    const memoryVectorStore = new MemoryVectorStore({
      vectors: [
        {
          id: 'vec1',
          content: 'I like pizza',
          embedding: await embed('I like pizza'),
          metadata: {},
        },
        {
          id: 'vec2',
          content: "I don't like dogs",
          embedding: await embed("I don't like dogs"),
          metadata: {},
        },
        {
          id: 'vec2',
          content: 'I looks like a clown',
          embedding: await embed('I looks like a clown'),
          metadata: {},
        },
      ],
    });

    const retriever = new VectorStoreRetriever({
      vectorstore: memoryVectorStore,
      search: {
        type: 'mmr',
        topK: 2,
        lambda: 0.5
      },
    });

    expect(
      await retriever.invoke(await embed('what do I like?'))
    ).toMatchSnapshot();
  });
});
