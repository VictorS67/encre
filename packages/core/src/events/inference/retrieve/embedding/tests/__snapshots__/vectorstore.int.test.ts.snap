// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VectorStoreRetriever mmr search 1`] = `
[
  Context {
    "metadata": {},
    "pageContent": "I like pizza",
  },
  Context {
    "metadata": {},
    "pageContent": "I don't like dogs",
  },
]
`;

exports[`VectorStoreRetriever similarity search 1`] = `
[
  Context {
    "metadata": {},
    "pageContent": "I like pizza",
  },
  Context {
    "metadata": {},
    "pageContent": "I don't like dogs",
  },
]
`;
